pkg/applicationserver/applicationserver.go

```diff
@@ -1352,6 +1412,28 @@ func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids *ttnpb.
 			}
 
 			items := []*ttnpb.ApplicationDownlink{msg}
+
+			if msg.Confirmed {
```

Comment:
I think this `if` is redundant - it is not possible (unless a bug occurs) that `msg.Confirmed` is false if it arrives via a `NACK`.